src/core/CustomData.cpp

```diff
@@ -0,0 +1,152 @@
+/*
+ *  Copyright (C) 2018 KeePassXC Team <team@keepassxc.org>
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 2 or (at your option)
+ *  version 3 of the License.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "CustomData.h"
+
+CustomData::CustomData(QObject* parent)
+    : QObject(parent)
+{
+}
+
+QList<QString> CustomData::keys() const
+{
+    return m_data.keys();
+}
+
+bool CustomData::hasKey(const QString& key) const
+{
+    return m_data.contains(key);
+}
+
+QString CustomData::value(const QString& key) const
+{
+    return m_data.value(key);
+}
+
+bool CustomData::contains(const QString& key) const
+{
+    return m_data.contains(key);
+}
+
+bool CustomData::containsValue(const QString& value) const
+{
+    return m_data.values().contains(value);
+}
+
+void CustomData::set(const QString& key, const QString& value)
+{
+    bool addAttribute = !m_data.contains(key);
+    bool changeValue = !addAttribute && (m_data.value(key) != value);
+
+    if (addAttribute) {
+        emit aboutToBeAdded(key);
+     }
+
+    if (addAttribute || changeValue) {
+        m_data.insert(key, value);
+        emit modified();
+    }
+
+    if (addAttribute) {
+        emit added(key);
+    }
+}
+
+void CustomData::remove(const QString& key)
+{
+    emit aboutToBeRemoved(key);
+
+    m_data.remove(key);
+
+    emit removed(key);
+    emit modified();
+}
+
+void CustomData::rename(const QString& oldKey, const QString& newKey)
```

Comment:
This function is never used and I don't see a case for it either. Custom Data is not directly user editable. Recommend remove.

Reply:
It may be useful for plugins, but I don't know how often a plugin needs to rename a field.